utils/diffing_with_keys.ml

```diff
@@ -78,38 +91,74 @@ let add_edge ex ey = match ex, ey with
   | Both _ as b, _ | _, Some (Both _ as b)  -> Some b
   | l, _ -> Some l
 
-let exchanges ~update ~key state changes =
-  let add (state,(swaps,moves)) d =
+type ('l, 'r, 'state) swaps =
+  ('l with_pos * 'r with_pos, 'l with_pos * 'r with_pos, 'state)
+    partial_edge Swap.t
+type ('l, 'r, 'state) moves =
+  ('l with_pos, 'r with_pos, 'state)
+    partial_edge Move.t
+
+type ('l, 'r, 'eq, 'diff, 'state) update =
+  ('l with_pos, 'r with_pos, 'eq, 'diff) Diffing.change -> 'state -> 'state
+type ('l, 'r, 'eq, 'diff, 'state) test =
+  'state -> 'l with_pos -> 'r with_pos -> ('eq, 'diff) result
+
+let exchanges (type l r eq diff state)
+    ~(update : (l, r, eq, diff, state) update)
+    ~(key : (l, r) keys)
+    (state : state)
+    (changes : (l with_pos, r with_pos, eq, diff) Diffing.change list)
+  : state * ((l, r, state) swaps * (l, r, state) moves)
+=
+  let add (state,
+           ((swaps : (l, r, state) swaps),
+            (moves : (l, r, state) moves))) d =
     update d state,
     match d with
     | Diffing.Change (x,y,_) ->
-        let k, edge = edge key state x y in
+        let k, edge = swap_edge key state x y in
         Swap.update k (add_edge edge) swaps, moves
     | Diffing.Insert nx ->
-        let k = key (data nx) in
+        let k = snd key (data nx) in
         let edge = Right (pos nx, state,nx) in
         swaps, Move.update k (add_edge edge) moves
     | Diffing.Delete nx ->
-        let k, edge = key (data nx), Left (pos nx, state, nx) in
+        let k, edge = fst key (data nx), Left (pos nx, state, nx) in
         swaps, Move.update k (add_edge edge) moves
     | _ -> swaps, moves
   in
-  List.fold_left add (state,(Swap.empty,Move.empty)) changes
+  List.fold_left add (state,(Swap.empty, Move.empty)) changes
 
 
-let swap key test swaps x y =
-  let kx, ky = key (data x), key (data y) in
-  let key = if kx <= ky then kx, ky else ky, kx in
-  match Swap.find_opt key swaps with
+let swap (type l r state eq diff)
+    ~(key : (l, r) keys)
+    ~(test : (l, r, eq, diff, state) test)
+    (swaps : (l, r, state) swaps)
+    (x : l with_pos) (y : r with_pos)
+  : (string with_pos * string with_pos) option
+=
+  let kx, ky = fst key (data x), snd key (data y) in
+  let k = if kx <= ky then kx, ky else ky, kx in
+  match Swap.find_opt k swaps with
   | None | Some (Left _ | Right _)-> None
   | Some Both (state, (ll,lr),(rl,rr)) ->
-      match test state ll rr,  test state lr rl with
+      match test state ll rr, test state rl lr with
```

Comment:
There is a change here, that is mandated by the typing from the choice of using `(l * r, l * r) partial_edge` for swaps.

I believe that the change is not observable, due to the fact that the typeful information inside the `result` case is ignored (and likely to be uninteresting in the `Ok` case anyway, which is the one we care about here).